src/amqp.js

```diff
@@ -0,0 +1,251 @@
+    logger.log('debug', 'Ack and reply messages!');
+    messages.forEach((message, index) => {
+      const headers = getHeaderInfo(message);
+
+      // Reply consumer gets: {"data":{"status":"UPDATED","payload":"0"}}
+      sendToQueue({
+        queue: message.properties.correlationId,
+        correlationId: message.properties.correlationId,
+        headers,
+        data: {
+          status, payload: payloads[index]
+        }
+      });
+
+      channel.ack(message);
+    });
+  }
+
+  function ackMessages(messages) {
+    logger.log('debug', 'Ack messages!');
+    messages.forEach(message => {
+      channel.ack(message);
+    });
+  }
+
+  function nackMessages(messages) {
+    logger.log('debug', 'Nack messages!');
+    messages.forEach(message => {
+      // Message, allUpTo, reQueue
+      channel.reject(message, true);
+    });
+  }
+
+  async function sendToQueue({queue, correlationId, headers, data}) {
+    try {
+      // Logger.log('debug', `Record queue ${queue}`)
+      // Logger.log('debug', `Record correlationId ${correlationId}`);
+      // Logger.log('debug', `Record data ${data}`);
+      // Logger.log('debug', `Record headers ${headers}`);
+
+      await channel.assertQueue(queue, {durable: true});
+
+      channel.sendToQueue(
+        queue,
+        Buffer.from(JSON.stringify({data})),
+        {
+          correlationId,
+          persistent: true,
+          headers
+        }
+      );
+
+      // Spams: logger.log('debug', `Message send to queue ${queue}`);
+    } catch (error) {
+      logError(error);
+    }
+  }
+
+  async function removeQueue(queue) {
+    await channel.deleteQueue(queue);
+  }
+
+  // ----------------
+  // Helper functions
+  // ----------------
+
+  function messagesToRecords(messages) {
+    logger.log('debug', 'Parsing messages to records');
+
+    return messages.map(message => {
+      const content = JSON.parse(message.content.toString());
+      return new MarcRecord(content.data);
+    });
+  }
+
+  async function getData(queue) {
+    try {
+      const {messageCount} = await channel.checkQueue(queue);
+      console.log('messageCount', messageCount); // eslint-disable-line no-console
+      const messagesToGet = messageCount >= CHUNK_SIZE ? CHUNK_SIZE : messageCount;
+      console.log('messagesToGet', messagesToGet); // eslint-disable-line no-console
+
+      const messages = Array(messagesToGet).map(() => {
+        const message = channel.get(queue)
+        console.log('message', message);
+        return message;
+      });
+      console.log('messages', messages);
+      await Promise.all(messages);
+
+      console.log(JSON.stringify(messages)); // eslint-disable-line no-console
+      const uniqueMessages = messages.filter(onlyUniques);
+      console.log(JSON.stringify(uniqueMessages)); // eslint-disable-line no-console
+
+      return uniqueMessages;
+    } catch (error) {
+      logError(error);
+    }
+
+    function onlyUniques(value, index, self) {
```

Comment:
I don't think this does what you want. Assuming you want to return only one instance of duplicate message? You need to use reduce for this.

Reply:
it works it is safequard that allows only unique messages pass it. If some reason there is duplicate message it removes the later one.

Reply:
Got it. But don't use `indexOf` for this. It's hard to understand (Since you don't really want to get the index of the value, you want to know if the value exists. Also, this requires knowing that `indexOf` only returns the first index of the value). Using `reduce` makes this more explicit for filtering out duplicates (And ECMAscript doesn't yet implement such a function): `list.reduce((acc, v) => v in acc ? acc : acc.concat(v))`